crates/proc-macro-srv/src/abis/mod.rs

```diff
@@ -69,6 +75,14 @@ impl Abi {
         symbol_name: String,
         info: RustCInfo,
     ) -> Result<Abi, LoadProcMacroDylibError> {
+        // Gated behind an env var for now to avoid a change in behavior for
+        // rustup-installed rust-analyzer
+        #[cfg(feature = "sysroot-abi")]
+        if std::env::var("PROC_MACRO_SRV_SYSROOT_ABI").is_ok() {
```

Comment:
I think this should also check that the proc macro version matches the version with which rust-analyzer was compiled and fallback to the numbered sysroot versions otherwise. Otherwise enabling this option would break when mixing a nightly rust-analyzer with a stable rustc.

Reply:
> Otherwise enabling this option would break when mixing a nightly rust-analyzer with a stable rustc.

That's interesting and I'm probably going to end up doing this, but here's how I thought it would work:

In the short term, nothing changes, the sysroot ABI is never used. It's not built into vscode-marketplace builds of rust-analyzer, and it's behind that env var for rustup builds (but it is there).

In the longer term, rust-analyzer is smart enough to find out if, for a given proc macro lib: 1) it can use (or `rustup component add` then use) a rustup-provided rust-analyzer binary 2) that binary supports sysroot-abi. And then it just turns on the sysroot ABI.

If different rust toolchains are used from the same, uh, "vscode window", it would just use different versions of the `rust-analyzer` binary, each with their own sysroot ABI matching the version used to compile the proc macro.

**This is just for the proc macro server** - this does _not_ tie rust-analyzer's feature release cycle to rust stable.

Reply:
That said, I like your idea and I'm going to add it anyway, because it's a cool easy fallback with basically zero downsides.